### ISSUE TITLE
Startup race leads to unmanaged pods in kubernetes 

### ISSUE THREAD

**Issue by opaetzel** [cilium/cilium#23394]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We deployed cilium in CNI chaining mode with AWS CNI. We seem to have a race when a new node comes up. Sometimes pods from other daemonsets (like opentelemetry-collector) will not be managed by cilium, e.g. they don't show up in cilium endpoint list . We are already using the taint `  node.cilium.io/agent-not-ready=true:NoSchedule`

The attached kubelet logs show the startup of a node where an opentelemetry-collector pod came up unmanaged (node_up_kubelet_logs_broken.log) and the kubelet logs of a node where opentelemetry-collector came up managed by cilium (node_up_kubelet_logs_ok.log).  In the case where we ended up with an unmanaged pod, opentelemetry-collector is started within the same second cilium came up. On the node where opentelemetry got managed by cilium, there are 3 seconds between cilium being up and opentelemetry being started.
[node_up_kubelet_logs_broken.log](https://github.com/cilium/cilium/files/10511699/node_up_kubelet_logs_broken.log)
[node_up_kubelet_logs_ok.log](https://github.com/cilium/cilium/files/10511700/node_up_kubelet_logs_ok.log)

### Cilium Version

v1.12.5

### Kernel Version

5.15.0-1026-aws #30~20.04.2-Ubuntu SMP Fri Nov 25 14:53:22 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.22.17

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aanm** [comment]
@opaetzel do you have cilium-operator logs that show when the taint was removed?

**Comment by opaetzel** [comment]
I couldn't find anything in the cilium-operator logs regarding the taint. But I had prepared a script to watch the node resources, so I now have timings for a new node that came up and had the same problem. 
The kubelet reported the cilium agent to be ready at 07:38:22.768229, whereas the taint seems to have been removed earlier, according to the watch I did for the node resources:

```
{"time": "2023-01-27 07:38:18.537610", "node": "ip-172-20-38-56.eu-central-1.compute.internal", "taints": [{"effect": "NoSchedule", "key": "node.cilium.io/agent-not-ready", "value": "true"}, {"effect": "NoExecute", "key": "node.kubernetes.io/not-ready", "timeAdded": "2023-01-27T07:37:24+00:00"}]}
{"time": "2023-01-27 07:38:20.321253", "node": "ip-172-20-38-56.eu-central-1.compute.internal", "taints": [{"effect": "NoExecute", "key": "node.kubernetes.io/not-ready", "timeAdded": "2023-01-27T07:37:24+00:00"}]}
```


[new_node_problem_kubelet.log](https://github.com/cilium/cilium/files/10518581/new_node_problem.log)

**Comment by squeed** [comment]
https://github.com/cilium/cilium/pull/23486 should improve matters significantly. It is, however, a rewrite rather than a bugfix, so I'm not sure if we can backport any of it.

**Comment by christarazi** [comment]
@opaetzel Would you be able to try https://github.com/cilium/cilium/pull/23486?

**Comment by opaetzel** [comment]
@christarazi I am not sure how I would do that. I am deploying cilium with the helm chart. Is there an image I could use or would I need to compile cilium myself? Also which images would I need to change? I would guess agent and operator, but I am not sure.

**Comment by christarazi** [comment]
Yes you can [build](https://docs.cilium.io/en/v1.12/contributing/development/images/#developer-images) Cilium with the changes checked out locally. Then you would deploy Cilium (Agent and Operator) with the helm flags set to the images newly built.

**Comment by christarazi** [comment]
@opaetzel Is still still an issue? Have you tried the above?

**Comment by opaetzel** [comment]
Hi, I have not yet tried the above. I will try with 1.14 when it comes out I guess. Maybe with the snapshot release for 1.14. I will update here, I have not forgotten the issue.

**Comment by opaetzel** [comment]
Just wanted to comment that it looks like this was fixed by installing cilium 1.14.1. I will open a new issue if this comes up again I guess.
